File: chess.cpp
#include <iostream>
#include <optional>
using namespace std;

enum class Color {
  black, white
};
string to_string(const Color& c) {
  if (c == Color::black)
    return "b";
  return "w";
}
enum class ptype {
  King, Queen
};
string to_string(const ptype& p) {
  switch (p) {
    case ptype::King:
      return "K";
    case ptype::Queen:
      return "Q";
    default:
      return "?";
  }
}
using Piece = optional<pair<Color, ptype>>;

string to_string(const Piece& p) {
  if (!p.has_value())
    return "-";
  return to_string(p.value().first) + to_string(p.value().second);
}

std::ostream& operator<<(std::ostream& os, const Piece& p) {
  os << to_string(p);
  return os;
}
const int board_size{8};
using loc = pair<int, int>;
loc left(const loc& l) {
  return loc(l.first, l.second - 1);
}
loc right(const loc& l) {
  return loc(l.first, l.second + 1);
}
loc up(const loc& l) {
  return loc(l.first - 1, l.second);
}
loc down(const loc& l) {
  return loc(l.first + 1, l.second);
}
class Board {
  public:
    Piece get_val(const loc& l) const {
      return board[l.first][l.second];
    }
    bool set_val(const loc& l, const Piece& p) {
      board[l.first][l.second] = p;
      return true;
    }
    void print() {
      for (int i = 0; i < board_size; ++i) {
        cout << to_string(i + 1) + "\t";
        for (int j = 0; j < board_size; ++j) {
          cout << board[i][j] << "\t";
        }
        cout << endl << endl;
      }
      cout << "\t";
      for (int i = 0; i < board_size; ++i) {
        cout << static_cast<char>('a' + i) << "\t";
      }
      cout << endl;
    }
  private:
    Piece board[board_size][board_size];
};
void init_board(Board& board) {
  auto wk = make_pair(Color::white, ptype::King);
  auto bk = make_pair(Color::black, ptype::King);
  auto wq = make_pair(Color::white, ptype::Queen);
  auto bq = make_pair(Color::black, ptype::Queen);
  board.set_val(make_pair(0, 0),wk);
  board.set_val(make_pair(0, 1),wq);
  board.set_val(make_pair(board_size - 1, board_size - 1),bk);
  board.set_val(make_pair(board_size - 1, board_size - 2),bq);
}
auto check_value = [](const int& val) -> bool {
    return 0 <= val && val < board_size;
  };
auto in_boundary = [](const loc& l) -> bool {
  return check_value(l.first) && check_value(l.second);
};

bool visit[board_size][board_size];
void init_visit() {
  for (size_t i = 0; i < board_size; ++i) {
    for (size_t j = 0; j < board_size; ++j) {
      visit[i][j] = false;
    }
  }
}
bool is_empty_slot(const Board& b, const loc& l) {
  return b.get_val(l) == nullopt;
}

bool queen_reachable(const Board& b, const loc& cur, const loc& dest) {
  if (!in_boundary(cur) || !is_empty_slot(b, cur) ||
      visit[cur.first][cur.second]) {
    return false;
  }
  if (cur == dest)
    return true;
  else {
    visit[cur.first][cur.second] = true;
    //left, right, up, down, leftup, leftdown, rightup, rightdown
    return queen_reachable(b, left(cur), dest) ||
          queen_reachable(b, right(cur), dest) ||
          queen_reachable(b, up(cur), dest) || 
          queen_reachable(b, down(cur), dest) ||
          queen_reachable(b, up(left(cur)), dest) ||
          queen_reachable(b, up(right(cur)), dest) ||
          queen_reachable(b, down(left(cur)), dest) ||
          queen_reachable(b, down(right(cur)), dest);
  }
}
bool king_reachable(const Board& b, const loc& cur, const loc& dest) {
  return abs(cur.first - dest.first) <= 1 &&
        abs(cur.second - dest.second) <= 1;
}
bool reachable(const Board& b, const loc& l1, const loc& l2) {
  init_visit();
  auto p = b.get_val(l1);
  const auto& piece_type = p.value().second;
  if (piece_type == ptype::Queen) {
    return queen_reachable(b, l1, l2);
  } else if (piece_type == ptype::King) {
    return king_reachable(b, l1, l2);
  } else {
    return false;
  }
}

void print_menu(const int& turn) {
  cout << "turn " << turn << endl;
  cout << "press q to quit" << endl;
  cout << "input example : a1 b1" << endl;
}
string to_string(const loc& l) {
  return "[" + to_string(l.first) + "][" + to_string(l.second) + "]";
}
// check if input is valid
// 1. l1 & l2 should be in boundary
// 2. turn & piece color should be matched.
bool validate_input(const Board& b, int turn, const loc& l1, const loc& l2) {
  // location should be in board boundary
  if (!in_boundary(l1) || !in_boundary(l2)) {
    cout << "boundary error\n";
    cout << to_string(l1) << ", " << to_string(l2) << endl;
    return false;
  }
  // remain same area is false
  if (l1 == l2) {
    cout << "same area error\n";
    return false;
  }

  const auto& current_val = b.get_val(l1);

  // empty area
  if (!current_val.has_value()) {
    cout << "empty area error\n";
    return false;
  }
  // even turn = black's turn
  if (turn % 2 == 0) {
    if (current_val.value().first == Color::white) {
      cout << "turn error\n";
      return false;
    }
  } else {
    // odd turn = white's turn
    if (current_val.value().first == Color::black) {
      cout << "turn error\n";
      return false;
    }
  }
  return true;
}
bool process_move(Board& b, const loc& l1, const loc& l2) {
  init_visit();
  auto p1 = b.get_val(l1);
  auto p2 = b.get_val(l2);
  if (p2.has_value() && p2.value().second == ptype::King) {
    cout << to_string(p1.value().first) << " wins!\n";
    return true;
  }
  b.set_val(l2, b.get_val(l1).value());
  b.set_val(l1, nullopt);
  return false;
}
void play(Board& b, int turn) {
  // print in every turn
  print_menu(turn);
  b.print();

  // input
  string input1{}, input2;
  cin >> input1;
  if (input1[0] == 'q')
    return;
  cin >> input2;
  int col1 = input1[0] - 'a';
  int row1 = input1[1] - '1';
  int col2 = input2[0] - 'a';
  int row2 = input2[1] - '1';
  auto l1 = make_pair(row1, col1);
  auto l2 = make_pair(row2, col2);
  if (!validate_input(b, turn, l1, l2)) {
    cout << "Invalid input. Please reinput command.\n";
  } else {
    if (process_move(b, l1, l2))
      return;
    turn += 1;
  }
  return play(b, turn);
}
int main() {
  Board b{};
  init_board(b);
  play(b, 1);
}

File: chess.py
from enum import Enum


def composite_function(f, g):
    return lambda x: f(g(x))


def double(a):
    return 2 * a


def plus_one(b):
    return  b + 1


class Color(Enum):
    BLACK = "b"
    WHITE = "w"


class PieceType(Enum):
    KING = "K"
    QUEEN = "Q"
    BISHOP = "B"
    KNIGHT = "N"
    ROOK = "R"
    PAWN = "P"



def left(loc : tuple):
    return loc[0], loc[1] - 1


def right(loc):
    return loc[0], loc[1] + 1


def up(loc):
    return loc[0] - 1, loc[1]


def down(loc):
    return loc[0] + 1, loc[1]


board_size = 8
board = []


def init_board():
    wk = Color.WHITE, PieceType.KING
    wq = Color.WHITE, PieceType.QUEEN
    wb1 = Color.WHITE, PieceType.BISHOP
    wb2 = Color.WHITE, PieceType.BISHOP
    wr1 = Color.WHITE, PieceType.ROOK
    wr2 = Color.WHITE, PieceType.ROOK
    wn1 = Color.WHITE, PieceType.KNIGHT
    wn2 = Color.WHITE, PieceType.KNIGHT
    bk = Color.BLACK, PieceType.KING
    bq = Color.BLACK, PieceType.QUEEN
    bb1 = Color.BLACK, PieceType.BISHOP
    bb2 = Color.BLACK, PieceType.BISHOP
    br1 = Color.BLACK, PieceType.ROOK
    br2 = Color.BLACK, PieceType.ROOK
    bn1 = Color.BLACK, PieceType.KNIGHT
    bn2 = Color.BLACK, PieceType.KNIGHT
    white_pawn = [Color.WHITE, PieceType.PAWN, False]
    wp_row = [white_pawn[:] for _ in range(board_size)]
    black_pawn = [Color.BLACK, PieceType.PAWN, False]
    bp_row = [black_pawn[:] for _ in range(board_size)]
    black_row = [br1, bn1, bb1, bk, bq, bb2, bn2, br2]
    white_row = [wr1, wn2, wb1, wk, wq, wb2, wn2, wr2]
    board.append(black_row)
    board.append(bp_row)
    for i in range(4):
        board.append([None] * board_size)

    board.append(wp_row)
    board.append(white_row)


def print_board():
    for i in range(board_size):
        print(i + 1, end='\t')
        for j in range(board_size):
            p = board[i][j]
            if p is None:
                print('-', end='\t')
            else:
                print(str(p[0].value) + str(p[1].value), end='\t')
        print()
    colphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
    print("", end='\t')
    print(*colphabet, sep='\t')


def check_value(val):
    return 0 <= val < board_size


def in_boundary(loc):
    return check_value(loc[0]) and check_value(loc[1])


def recursive_reachable(cur, dest, func) -> bool:
    if cur == dest:
        return True

    if in_boundary(cur) is False or get_val(cur) is not None:
        return False
    else:
        return recursive_reachable(func(cur), dest, func)


def king_reachable(cur, dest):
    val1 = abs(cur[0] - dest[0])
    val2 = abs(cur[1] - dest[1])
    return val1 <= 1 and val2 <= 1 and not (val1 == 0 and val2 == 0)


def knight_reachable(cur, dest):
    return up(up(left(cur))) == dest or \
            up(up(right(cur))) == dest or \
            down(down(left(cur))) == dest or \
            down(down(right(cur))) == dest or \
            right(right(up(cur))) == dest or \
            right(right(down(cur))) == dest or \
            left(left(up(cur))) == dest or \
            left(left(down(cur))) == dest


def pawn_reachable(cur, dest):
    def small_reachable(func):
        if func(left(cur)) == dest and get_val(dest) is not None:
            return True
        elif func(right(cur)) == dest and get_val(dest) is not None:
            return True
        elif func(cur) == dest:
            return True
        elif func(func(cur)) == dest:
            return get_val(func(cur)) is None and get_val(cur)[2] is False

        return False
    if get_val(cur)[0] == Color.WHITE:
        return small_reachable(up)
    else:
        return small_reachable(down)


def get_val(loc):
    return board[loc[0]][loc[1]]


def reachable(l1, l2) -> bool:
    p = get_val(l1)
    ptype = p[1]
    if ptype == PieceType.QUEEN:
        return recursive_reachable(left(l1), l2, left) or \
                recursive_reachable(down(l1), l2, down) or \
                recursive_reachable(up(l1), l2, up) or \
                recursive_reachable(right(l1), l2, right) or \
                recursive_reachable(up(right(l1)), l2, composite_function(up, right)) or \
                recursive_reachable(up(left(l1)), l2, composite_function(up, left)) or \
                recursive_reachable(down(right(l1)), l2, composite_function(down, right)) or \
                recursive_reachable(down(left(l1)), l2, composite_function(down, left))
    elif ptype == PieceType.KING:
        return king_reachable(l1, l2)
    elif ptype == PieceType.BISHOP:
        return recursive_reachable(up(right(l1)), l2, composite_function(up, right)) or \
                recursive_reachable(up(left(l1)), l2, composite_function(up, left)) or \
                recursive_reachable(down(right(l1)), l2, composite_function(down, right)) or \
                recursive_reachable(down(left(l1)), l2, composite_function(down, left))
    elif ptype == PieceType.ROOK:
        return recursive_reachable(left(l1), l2, left) or \
               recursive_reachable(down(l1), l2, down) or \
               recursive_reachable(up(l1), l2, up) or \
               recursive_reachable(right(l1), l2, right)
    elif ptype == PieceType.KNIGHT:
        return knight_reachable(l1, l2)
    elif ptype == PieceType.PAWN:
        return pawn_reachable(l1, l2)
    else:
        return False


def print_menu(turn):
    print("turn " + str(turn))
    print("press q to quit")
    print("input example : b1 h7")


# check if input is valid
# 1. l1 & l2 should be in boundary
# 2. turn & piece color should be matched.
def validate_input(turn, l1, l2):
    # location should be in board boundary
    if in_boundary(l1) is False or in_boundary(l2) is False:
        print("boundary error")
        print(l1 + ", " + l2)
        return False
    # remain same spot is false
    if l1 == l2:
        print("same spot error")
        return False

    cur_piece = get_val(l1)
    # empty spot
    if cur_piece is None:
        print("empty spot error")
        return False
    # color error
    dest_piece = get_val(l2)
    if dest_piece is not None:
        if cur_piece[0] == dest_piece[0]:
            print("same color error")
            return False
    # even turn = black's turn
    if turn % 2 == 0:
        if cur_piece[0] == Color.WHITE:
            print("turn error")
            return False
    elif cur_piece[0] == Color.BLACK:
        # odd turn = white's turn
        print("turn error")
        return False
    # check if reachable
    if reachable(l1, l2) is False:
        print("unreachable error")
        return False

    return True


def process_move(l1, l2):
    piece1 = get_val(l1)
    piece2 = get_val(l2)
    if piece2 is not None and piece2[1] == PieceType.KING:
        print(str(piece1[0]) + " wins!\n")
        return True

    if piece1[1] == PieceType.PAWN:
        piece1[2] = True

    board[l2[0]][l2[1]] = piece1
    board[l1[0]][l1[1]] = None
    return False


def play(turn):
    # print in every turn
    print_menu(turn)
    print_board()

    # input
    my_input = input()
    if my_input[0] == 'q':
        return
    in1, in2 = my_input.split(' ')
    col1 = ord(in1[0]) - ord('a')
    row1 = int(in1[1]) - 1
    col2 = ord(in2[0]) - ord('a')
    row2 = int(in2[1]) - 1
    l1 = row1, col1
    l2 = row2, col2
    if validate_input(turn, l1, l2) is False:
        print("Invalid input. Please reinput command.")
    else:
        if process_move(l1, l2) is True:
            return
        turn += 1

    return play(turn)


init_board()
play(1)
